pkg/yurthub/filter/filter.go

```diff
@@ -115,89 +116,151 @@ func (fis FilterInitializers) Initialize(ins Runner) error {
-		serializer: serializer,
-		ownerName:  ownerName,
-		stopCh:     stopCh,
-	}
-
-	if dr.isWatch {
-		go func(req *http.Request, rc io.ReadCloser, ch chan watch.Event) {
-			err := handler.StreamResponseFilter(rc, ch)
+	respContentType, _ := util.RespContentTypeFrom(ctx)
+	s := CreateSerializer(respContentType, info, sm)
+	if s == nil {
+		klog.Errorf("skip filter, failed to create serializer in %s", ownerName)
+		return 0, rc, nil
+	}
+
+	frc := &filterReadCloser{
+		rc:          rc,
+		watchDataCh: make(chan *bytes.Buffer),
+		filterCache: new(bytes.Buffer),
+		serializer:  s,
+		handler:     handler,
+		isWatch:     info.Verb == "watch",
+		ownerName:   ownerName,
+		stopCh:      stopCh,
+	}
+
+	if frc.isWatch {
+		go func(req *http.Request, rc io.ReadCloser, ch chan *bytes.Buffer) {
+			err := frc.StreamResponseFilter(rc, ch)
 			if err != nil && err != io.EOF && !errors.Is(err, context.Canceled) {
-				klog.Errorf("filter(%s) watch response ended with error, %v", dr.ownerName, err)
+				klog.Errorf("filter(%s) watch response ended with error, %v", frc.ownerName, err)
 			}
-		}(req, rc, dr.ch)
-		return 0, dr, nil
+		}(req, rc, frc.watchDataCh)
+		return 0, frc, nil
 	} else {
-		var newData []byte
-		n, err := dr.data.ReadFrom(rc)
-		if err != nil {
-			return int(n), dr, err
-		}
-
-		newData, err = handler.ObjectResponseFilter(dr.data.Bytes())
-		dr.data = bytes.NewBuffer(newData)
-		return len(newData), dr, err
+		var err error
+		frc.filterCache, err = frc.ObjectResponseFilter(rc)
+		return frc.filterCache.Len(), frc, err
 	}
 }
 
-// Read read data into p and write into pipe
-func (dr *filterReadCloser) Read(p []byte) (int, error) {
-	if dr.isWatch {
+// Read get data into p and write into pipe
+func (frc *filterReadCloser) Read(p []byte) (int, error) {
+	var ok bool
+	if frc.isWatch {
+		if frc.filterCache.Len() != 0 {
+			return frc.filterCache.Read(p)
+		} else {
+			frc.filterCache.Reset()
+		}
+
 		select {
-		case watchEvent, ok := <-dr.ch:
+		case frc.filterCache, ok = <-frc.watchDataCh:
 			if !ok {
 				return 0, io.EOF
 			}
+			return frc.filterCache.Read(p)
+		}
+	} else {
+		return frc.filterCache.Read(p)
+	}
+}
 
-			buf := &bytes.Buffer{}
-			n, err := dr.serializer.WatchEncode(buf, &watchEvent)
-			if err != nil {
-				klog.Errorf("filter(%s) failed to encode resource in Reader %v", dr.ownerName, err)
-				return 0, err
-			}
-			copied := copy(p, buf.Bytes())
-			if copied != n {
-				return 0, fmt.Errorf("filter(%s) expect copy %d bytes, but only %d bytes copyied", dr.ownerName, n, copied)
-			}
+// Close will close readers
+func (frc *filterReadCloser) Close() error {
+	if frc.filterCache != nil {
+		frc.filterCache.Reset()
+	}
+	return frc.rc.Close()
+}
+
+func (frc *filterReadCloser) ObjectResponseFilter(rc io.ReadCloser) (*bytes.Buffer, error) {
+	var buf bytes.Buffer
+	_, err := buf.ReadFrom(rc)
+	if err != nil {
+		return &buf, err
+	}
+	listObj, err := frc.serializer.Decode(buf.Bytes())
```

Comment:
It seems that we do not only handle listobject but also standalone object. So, it maybe unsuitable to use the variable name as `listObj`.

Reply:
listObj --> obj. 
fixed